### PR TITLE
add static adamw

### PR DESCRIPTION
**Pull request by zhaoyinglia**
### PR types

New features
### PR changes

OPs 
### Describe

Add an optimizer for static graph

### CODE REVIEW

File: python/paddle/fluid/optimizer.py [hunk 1]
```diff
@@ -2561,6 +2561,261 @@ def _finish_update(self, block, parameters_and_grads):
...
+            gradient in current mini-batch, so it will be much more faster. But this mode has
+            different semantics with the original Adam algorithm and may lead to different result.
+            The default value is False.
+        use_global_beta_pow (bool, optional): Whether to use global beta_pow. If true, Adam will use global beta_pow 
+            for whole model instead of creating beta_pow for each parameter. Default is false.
+        flatten_param_grads (bool, optional): Whether to flatten all parameters and gradients. Default is false.
+        align_size (int, optional): The alignment size when flatten parameters and gradients. Default is -1, which means
+            use same align_size as allocator.
+    
+    """
```
Comment (JZ-LIANG) [inline python/paddle/fluid/optimizer.py]: should add example here.

File: python/paddle/fluid/optimizer.py [hunk 2]
```diff
@@ -2561,6 +2561,261 @@ def _finish_update(self, block, parameters_and_grads):
...
+            raise ValueError("Invaild value of beta2, expect beta1 in [0,1).")
+        if not 0 <= epsilon:
+            raise ValueError("Invaild value of epsilon, expect epsilon >= 0.")
+        if not 0 <= weight_decay:
+            raise ValueError(
+                "Invaild value of weight_decay, expect weight_decay >= 0.")
+        if not isinstance(weight_decay, float) and \
+                not isinstance(weight_decay, framework.Variable):
+            raise TypeError("weight_decay should be float or Tensor.")
+
```
Comment (JZ-LIANG) [inline python/paddle/fluid/optimizer.py]: should check if in dygraph mode or not here, and raise NotImplementedError if it is in dygraph mode. 
you could find an example in RecomputeOptimizer

File: paddle/fluid/operators/optimizers/adamw_op.cc [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+  PADDLE_ENFORCE_EQ(ctx->HasOutput("ParamOut"), true,
+                    platform::errors::NotFound(
+                        "Output(ParamOut) of AdamWOp should not be null."));
+  PADDLE_ENFORCE_EQ(ctx->HasOutput("Moment1Out"), true,
+                    platform::errors::NotFound(
+                        "Output(Moment1Out) of AdamWOp should not be null."));
+  PADDLE_ENFORCE_EQ(ctx->HasOutput("Moment2Out"), true,
+                    platform::errors::NotFound(
+                        "Output(Moment2Out) of AdamWOp should not be null."));
+
```
Comment (JZ-LIANG) [inline paddle/fluid/operators/optimizers/adamw_op.cc:55]: should also checking weight_decay here

File: paddle/fluid/operators/optimizers/adamw_op.cc [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+             "(Tensor<float32>, optional) If provided, AdamW will use this "
+             "as beta2, this has a higher priority than attr(beta2), the "
+             "shape of this tensor MUST BE [1].")
+        .AsDispensable();
+    AddInput("EpsilonTensor",
+             "(Tensor<float32>, optional) If provided, AdamW will use this "
+             "as epsilon, this has a higher priority than attr(epsilon), the "
+             "shape of this tensor MUST BE [1].")
+        .AsDispensable();
+    AddInput("WeightDecayTensor",
```
Comment (JZ-LIANG) [inline paddle/fluid/operators/optimizers/adamw_op.cc:159]: it is OK to have "WeightDecayTensor", but it is not need to following "Epsilon" to have EpsilonTensor as well as Epsilon attribute two duplicated args for just one single logic.  not need to modify.

File: paddle/fluid/operators/optimizers/adamw_op.cc [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+    AddAttr<float>("beta2",
+                   "(float, default 0.999) "
+                   "exponential decay rate for the "
+                   "second moment estimates.")
+        .SetDefault(0.999f);
+    AddAttr<float>("epsilon",
+                   "(float, default 1.0e-8) "
+                   "Constant for numerical stability")
+        .SetDefault(1.0e-8f);
+    AddAttr<float>("weightdecay",
```
Comment (JZ-LIANG) [inline paddle/fluid/operators/optimizers/adamw_op.cc]: varname "weightdecay" should be "weight_decay" to follow the same pattern as in "lazy_mode"/ "multi_precision"

File: python/paddle/fluid/tests/unittests/test_adamw_op_static.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+from __future__ import print_function
+
+import unittest
+import numpy as np
+from paddle.fluid import core
+import paddle.fluid as fluid
+import paddle
+
+
+class testAdamWOpStatic(unittest.TestCase):
```
Comment (JZ-LIANG) [inline python/paddle/fluid/tests/unittests/test_adamw_op_static.py:24]: the unitest is tooooo simple !!!
you should take python/paddle/fluid/tests/unittests/test_adam_op.py as a reference, and AT LEAST has a unitest for accuracy. for example build a simple small network + feed same data, then compare the 5th-step-loss with paddle.optimizer.adamw, which should be exactly the same.

File: python/paddle/fluid/optimizer.py (hunk 1 above)
Comment (zhaoyinglia) [reply python/paddle/fluid/optimizer.py]: Ok, the example has been added.

File: python/paddle/fluid/optimizer.py (hunk 2 above)
Comment (zhaoyinglia) [reply python/paddle/fluid/optimizer.py]: It has been refined.

File: paddle/fluid/operators/optimizers/adamw_op.cc (hunk 3 above)
Comment (zhaoyinglia) [reply paddle/fluid/operators/optimizers/adamw_op.cc:55]: weight_decay variable has "WeightDecayTensor" and "weight_decay" attribution, so "WeightDecayTensor" input maybe not exist here.

File: paddle/fluid/operators/optimizers/adamw_op.cc (hunk 4 above)
Comment (zhaoyinglia) [reply paddle/fluid/operators/optimizers/adamw_op.cc:159]: done.

File: paddle/fluid/operators/optimizers/adamw_op.cc (hunk 5 above)
Comment (zhaoyinglia) [reply paddle/fluid/operators/optimizers/adamw_op.cc]: Ok, it has been refined.

File: python/paddle/fluid/tests/unittests/test_adamw_op_static.py (hunk 6 above)
Comment (zhaoyinglia) [reply python/paddle/fluid/tests/unittests/test_adamw_op_static.py:24]: A unittest for accuracy has been added.

**Review by JZ-LIANG — DISMISSED** [review]
LGTM

**Review by PangHua — APPROVED** [review]
LGTM

**Pull request closed without merging**